src/action/ActionManager.js

```diff
@@ -0,0 +1,301 @@
+     */
+    this._filters = new Map();
+
+    /**
+     * name -> {@link ActionMap}
+     * @type {Map<string, ActionMap>}
+     */
+    this._actionMaps = new Map();
+
+    /**
+     * zero or more {@link ActionMap}s may be active at any time
+     * @type {Map<string, ActionMap>}
+     */
+    this._activeActionMaps = new Map();
+
+    /**
+     * actionPath -> { action parameters }
+     * @type {Map<string, Object>}
+     */
+    this._activeActionInfos = new Map();
+
+    // Used during updates
+    this._activations = new Map();
+    this._deactivations = new Map();
+
+    if (loadDefaults) this.loadDefaults();
+  }
+
+  /*
+  @param inputPath {string} a full semantic path for an input, like '/input/keyboard/key/32'
+  @return [inputValue, inputSource]
+  */
+  queryInputPath(inputPath) {
+    for (let inputSourceInfo of this.inputSources) {
+      if (inputPath.startsWith(inputSourceInfo[0]) === false) {
+        continue;
+      }
+      return [inputSourceInfo[1].queryInputPath(inputPath.substring(inputSourceInfo[0].length)), inputSourceInfo[1]];
+    }
+    return [null, null];
+  }
+
+  /**
+   * Poll the input states and update Action state.
+   * If you're rendering inside a requestAnimationFrame response, call this at the top of your render loop.
+   */
+  poll() {
+    // Poll inputs
+    this._inputSources.forEach((inputSource, inputPath) => {
+      inputSource.poll();
+    });
+
+    this._activations.clear();
+    this._deactivations.clear();
+
+    this._activeActionMaps.forEach((actionMap, name) => {
+      actionMap.update(this.queryInputPath, (actionPath, active, actionParameters, inputSource) => {
+        if (active) {
+          // Accept the first activation
+          if (this._activations.has(actionPath)) return;
+          this._activations.set(actionPath, [actionPath, actionParameters, inputSource]);
+        } else {
+          this._deactivations.set(actionPath, [actionPath, actionParameters, inputSource]);
+        }
+      });
+    });
+
+    for (const [actionPath, actionParameters, inputSource] of this._activations.values()) {
+      this._deactivations.delete(actionPath); // If we activate it in this update, don't deactivate it
+      if (this._activeActionInfos.has(actionPath)) {
+        this._activeActionInfos.set(actionPath, actionParameters);
+        continue; // Changing an already existing action does not trigger an event
+      }
+      this._activeActionInfos.set(actionPath, actionParameters);
+      this._notifyListeners(actionPath, true, actionParameters, inputSource);
+    }
+    for (const [actionPath, actionParameters, inputSource] of this._deactivations.values()) {
+      if (this._activeActionInfos.has(actionPath) === false) {
+        continue; // Action is not active, so no new event
+      }
+      this._activeActionInfos.delete(actionPath);
+      this._notifyListeners(actionPath, false, actionParameters, inputSource);
+    }
+  }
+
+  loadDefaults() {
```

Comment:
I'd put this default out from the core or if not out from the core at least out from the main class.